In [61]:
%load_ext autoreload
%autoreload 2

import numpy as np
from scipy.optimize import minimize
from scipy.stats import qmc
import pandas as pd
import sys
import time
from contextlib import redirect_stdout
from algorithms import TunnelingAlgorithm, MRSAlgorithm
from annealing import AnnealingTunnelingAlgorithm
import functions as fn
from algo_tester import Algorithm_Tester


The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


## The entire thing about the 1d example
## Tests showing the improvements make it better
## ideally some type of annealing

In [ ]:
np.random.seed(42)
algorithm_tester = Algorithm_Tester()
algorithm_tester.run_tests(AnnealingTunnelingAlgorithm, verbose=False, improved=True)

Test: Shubert 2D, Avg Minima Found: 14.35 out of 18, Failure Rate: 0.00% over 20 trials, Time per Trial: 1.318 seconds.
Test: Shubert 3D, Avg Minima Found: 2.80 out of 81, Failure Rate: 0.00% over 10 trials, Time per Trial: 0.344 seconds.


c:\Users\nicol\DS_Projects\Seminar_paper\python\sem_env\Lib\site-packages\numpy\_core\fromnumeric.py:3824: RuntimeWarning: Mean of empty slice
  return _methods._mean(a, axis=axis, dtype=dtype,
c:\Users\nicol\DS_Projects\Seminar_paper\python\sem_env\Lib\site-packages\numpy\_core\_methods.py:142: RuntimeWarning: invalid value encountered in scalar divide
  ret = ret.dtype.type(ret / rcount)


Test: Shubert 6D, Avg Minima Found: nan out of 4374, Failure Rate: 100.00% over 10 trials, Time per Trial: 0.227 seconds.
Test: Six Hump Camel, Avg Minima Found: 2.00 out of 2, Failure Rate: 0.00% over 30 trials, Time per Trial: 1.661 seconds.
Test: Altered Shubert 2D, Avg Minima Found: 1.00 out of 1, Failure Rate: 20.00% over 30 trials, Time per Trial: 0.322 seconds.
Test: Function 15 2D, Avg Minima Found: 1.00 out of 1, Failure Rate: 0.00% over 30 trials, Time per Trial: 3.119 seconds.
Test: Function 15 3D, Avg Minima Found: 1.25 out of 1, Failure Rate: 0.00% over 20 trials, Time per Trial: 2.819 seconds.
Test: Function 15 4D, Avg Minima Found: 1.90 out of 1, Failure Rate: 0.00% over 10 trials, Time per Trial: 3.271 seconds.
Test: Function 16 5D, Avg Minima Found: 1.00 out of 1, Failure Rate: 0.00% over 10 trials, Time per Trial: 5.597 seconds.
Test: Function 16 8D, Avg Minima Found: 1.00 out of 1, Failure Rate: 10.00% over 10 trials, Time per Trial: 5.706 seconds.
Test: Function 16 

Test: Shubert 2D, Avg Minima Found: 14.90 out of 18, Failure Rate: 0.00% over 50 trials, Time per Trial: 2.676 seconds.
Test: Shubert 3D, Avg Minima Found: 6.20 out of 81, Failure Rate: 0.00% over 10 trials, Time per Trial: 1.114 seconds.
Test: Shubert 6D, Avg Minima Found: 3.00 out of 4374, Failure Rate: 50.00% over 10 trials, Time per Trial: 0.861 seconds.
Test: Six Hump Camel, Avg Minima Found: 2.00 out of 2, Failure Rate: 0.00% over 30 trials, Time per Trial: 0.223 seconds.
Test: Altered Shubert 2D, Avg Minima Found: 1.00 out of 1, Failure Rate: 26.67% over 30 trials, Time per Trial: 0.662 seconds.
Test: Function 15 2D, Avg Minima Found: 1.00 out of 1, Failure Rate: 0.00% over 30 trials, Time per Trial: 0.351 seconds.
Test: Function 15 3D, Avg Minima Found: 1.00 out of 1, Failure Rate: 0.00% over 20 trials, Time per Trial: 0.500 seconds.
Test: Function 15 4D, Avg Minima Found: 1.00 out of 1, Failure Rate: 0.00% over 10 trials, Time per Trial: 0.678 seconds.
Test: Function 16 5D, Avg Minima Found: 1.00 out of 1, Failure Rate: 0.00% over 10 trials, Time per Trial: 0.889 seconds.
Test: Function 16 8D, Avg Minima Found: 1.00 out of 1, Failure Rate: 0.00% over 10 trials, Time per Trial: 1.742 seconds.
Test: Function 16 10D, Avg Minima Found: 1.00 out of 1, Failure Rate: 0.00% over 10 trials, Time per Trial: 2.530 seconds.
[{'function': 'Function 16 10D', 'avg_pct_minima_found': np.float64(0.1), 'total_minima': 1, 'failure_rate': 0.0, 'nr_trials': 10}]

In [ ]:
np.random.seed(42)
failure_counter = 0
for _ in range(1000):
    ta = TunnelingAlgorithm(fn.altered_cos, fn.altered_cos_grad, [[-10, 10]], verbose=False, fixed_xm=False)
    x_stars, f_star = ta.apply_algorithm(np.array([np.random.uniform(-10, 10)]))
    if len(x_stars) == 1:
        failure_counter += 1
    elif len(x_stars) not in [1, 2]:
        print(f"Unexpected number of minima found: {len(x_stars)} at points {x_stars} with value {f_star}")
        break 
print(failure_counter)

In [ ]:
lenx = 2
i = 0
xx = [np.array([-1])]
while lenx ==2 or xx[0] < 0:
    i += 1
    np.random.seed(i)
    ta = TunnelingAlgorithm(fn.altered_cos, fn.altered_cos_grad, [[-10, 10]], verbose=False, fixed_xm=False)
    xx, f_star = ta.apply_algorithm(np.array([np.random.uniform(-10, 10)]))
    lenx = len(xx)
print(f"The seed {i} causes the algorithm to find only one minimum at {xx} with value {f_star}.")
print("The algorithm will be run again with this seed and verbose output.\n")

np.random.seed(i)
ta = TunnelingAlgorithm(fn.altered_cos, fn.altered_cos_grad, [[-10, 10]], verbose=True, fixed_xm=False)
xx, f_star = ta.apply_algorithm(np.array([np.random.uniform(-10, 10)]))

In [60]:
for _ in range(50):
    ta = AnnealingTunnelingAlgorithm(fn.shubert, fn.shubert_grad, [[-10, 10], [-10, 10]], verbose=True, improved=True)
    with open("logs.txt", "w") as f:
        with redirect_stdout(f):
            xx, f_star = ta.apply_algorithm(np.array([np.random.uniform(-10, 10), np.random.uniform(-10, 10)]))
            print(xx)
    print(f"Annealing found {len(xx)} minima with value {f_star} at {xx}")
for _ in range(50):
    ta = TunnelingAlgorithm(fn.shubert, fn.shubert_grad, [[-10, 10], [-10, 10]], verbose=False, improved=True)
    with open("logs.txt", "w") as f:
        with redirect_stdout(f):
            xx, f_star = ta.apply_algorithm(np.array([np.random.uniform(-10, 10), np.random.uniform(-10, 10)]))
            print(xx)
    print(f"Regular found {len(xx)} minima with value {f_star} at {xx}")

Annealing found 14 minima with value -186.73090883102387 at [array([-7.70831374, -0.8003211 ]), array([-1.42512843, -0.80032109]), array([4.8580569, 5.4828642]), array([-1.42512843, -7.0835064 ]), array([ 5.4828642 , -1.42512842]), array([-0.80032108, -1.42512831]), array([-7.08350641, -7.70831374]), array([-7.70831374,  5.48286421]), array([ 4.85805688, -0.8003211 ]), array([-0.80032109, -7.70831373]), array([ 4.85805687, -7.08350641]), array([-7.08350642, -1.42512842]), array([ 5.48286421, -7.70831374]), array([-7.08350641,  4.85805688])]
Annealing found 17 minima with value -186.73090883102378 at [array([-7.70831374,  5.48286421]), array([-0.8003211 ,  4.85805688]), array([-1.42512841,  5.48286422]), array([4.85805693, 5.4828643 ]), array([5.48286421, 4.85805689]), array([ 5.48286421, -1.42512843]), array([-1.42512842, -7.08350647]), array([ 5.48286422, -7.7083137 ]), array([-7.7083143 , -7.08350661]), array([-1.42512843, -0.8003211 ]), array([-7.08350641,  4.85805688]), array([-0.8

In [56]:
ta = AnnealingTunnelingAlgorithm(fn.shubert, fn.shubert_grad, [[-10, 10], [-10, 10]], verbose=True, improved=True)
with open("logs.txt", "w") as f:
    with redirect_stdout(f):
        xx, f_star = ta.apply_algorithm(np.array([np.random.uniform(-10, 10), np.random.uniform(-10, 10)]))
        print(xx)
print(f"Annealing found {len(xx)} minima with value {f_star} at {xx}")

Annealing found 17 minima with value -186.73090883102384 at [array([ 5.48286421, -1.42512843]), array([-7.70831332, -0.80032129]), array([-7.70831374,  5.48286421]), array([-7.08350651,  4.85805707]), array([-1.42512878, -0.80032123]), array([-0.80032099,  4.85805696]), array([5.4828642 , 4.85805694]), array([ 5.48286401, -7.70831332]), array([-0.80032137, -7.70831384]), array([-0.80032125, -1.42512833]), array([ 4.85805685, -0.80032126]), array([-1.42512856, -7.08350651]), array([4.85805688, 5.48286414]), array([-7.08350648, -7.70831342]), array([-1.42512843,  5.48286421]), array([ 4.85805699, -7.08350669]), array([-7.08350671, -1.42512853])]
